addons/mail/static/src/messaging/entity/follower/follower.js

```diff
@@ -0,0 +1,212 @@
+            } else {
+                args.push([this.channel.id]);
+            }
+            await this.env.rpc({
+                model: this.followedThread.model,
+                method: 'message_unsubscribe',
+                args
+            });
+            this.delete();
+        }
+
+        /**
+         * @returns {integer}
+         */
+        get resId() {
+            if (this.partner) {
+                return this.partner.id;
+            }
+            if (this.channel) {
+                return this.channel.id;
+            }
+            return 0;
+        }
+
+        /**
+         * @returns {string}
+         */
+        get resModel() {
+            if (this.partner) {
+                return this.partner.model;
+            }
+            if (this.channel) {
+                return this.channel.model;
+            }
+            return '';
+        }
+
+        /**
+         * @param {mail.messaging.entity.FollowerSubtype} subtype
+         */
+        selectSubtype(subtype) {
+            if (!this.selectedSubtypes.includes(subtype)) {
+                this.link({ selectedSubtypes: subtype });
+            }
+        }
+
+        /**
+         * Show (editable) list of subtypes of this follower.
+         */
+        async showSubtypes() {
+            const subtypesData = await this.env.rpc({
+                route: '/mail/read_subscription_data',
+                params: { follower_id: this.id },
+            });
+            this.unlink({ subtypes: null });
+            for (const data of subtypesData) {
+                const subtype = this.env.entities.FollowerSubtype.insert(data);
+                this.link({ subtypes: subtype });
+                if (data.followed) {
+                    this.link({ selectedSubtypes: subtype });
+                } else {
+                    this.unlink({ selectedSubtypes: subtype });
+                }
+            }
+            this._subtypesListDialog = this.env.messaging.dialogManager.open('FollowerSubtypeList', {
+                follower: this,
+            });
+        }
+
+        /**
+         * @param {mail.messaging.entity.FollowerSubtype} subtype
+         */
+        unselectSubtype(subtype) {
+            if (this.selectedSubtypes.includes(subtype)) {
+                this.unlink({ selectedSubtypes: subtype });
+            }
+        }
+
+        /**
+         * Update server-side subscription of subtypes of this follower.
+         */
+        async updateSubtypes() {
+            if (this.selectedSubtypes.length === 0) {
+                this.remove();
+            } else {
+                const kwargs = {
+                    subtype_ids: this.selectedSubtypes.map(subtype => subtype.id),
+                };
+                if (this.partner) {
+                    kwargs.partner_ids = [this.partner.id];
+                } else {
+                    kwargs.channel_ids = [this.channel.id];
+                }
+                await this.env.rpc({
```

Comment:
Mmmhhh I don't understand why ? Because here the dialog will be closed so I think no flicker can happen here

Reply:
It's not about flicker here. The menu should probably not even close if the RPC fails, the user should not have the impression that it worked, and the local data should not reflect an incorrect state.

Reply:
I see, indeed it makes sense

Reply:
Mmh, I think the rpc should not be done before because the selected subtypes is not totally linked to the db state. So a DB state [selected, not selected, selected] could totally be associated to a selected subtypes state [selected, selected, selected].

But I'm ok that if the rpc failed it should not close, but how do I handle an error in rpc ?

Reply:
You need to check the API of the RPC and/or make some tests (for example adapt python to raise, close server to simulate timeout, ...). Either it is a returning a promise with a specific value on fail to conditionally check, or it is rejecting in which case it would be a try/catch with await.

I think we are going to have a lot more cases arising with the local data not being totally linked to the DB, and it might or might not be a bad idea to work around it.

This is a paradigm where we need to decide if we want the local state to be correct first, and updates are pushed to server later, or if the server must validate any change.

Imagine if something else is depending on the data we are talking about. Should the rest of the interface use the new data before the server says it is ok?

This is also the famous issue that we have with message post, and the problem of restoring the local data after noticing an RPC fail, which can become impossible to do if other changes have been done assuming the data change was ok in the first place (eg. typing a new message, which one to keep then?).

I think it's much safer and much easier to handle if the local state is never mixed with the state validated with the server. If a data can become sync with the server, it should never change its value before a successful RPC and/or longpolling notif that says it is OK to change.

Reply:
Yeah I totally understand, but in this case I think the impact is very limited to the dialog. What would you do here ? Given the fact that the dialog is intended to be changed by the user and then sync with server when click on apply. If we click on cancel those are not synced. Given also the fact that on each showing of the dialog we reload from server and the only dependent data are shown in that dialog.

Reply:
We can keep it like this in this PR, but I would like to make a pass just after where we fix the issue on a separate commit to see what kind of diff is necessary for it.

I think the solution should be that the dialog should use a temporary state (local state or probably entity) that has the server value as default value, but can be changed separately. Then if we close/cancel we delete those temp data, and if we save we wait on the RPC and copy them over to the actual data.

Reply:
https://www.odoo.com/web#id=2244616&action=3206&active_id=1519&model=project.task&view_type=form&cids=1&menu_id=4720